sound/soc/sof/topology.c

```diff
@@ -2208,32 +2102,16 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	if (ret != 0) {
 		dev_err(scomp->dev, "error: parse process.cfg tokens failed %d\n",
 			le32_to_cpu(private->size));
-		goto err;
+		kfree(process);
+		return ret;
 	}
 
 	sof_dbg_comp_config(scomp, &process->config);
 
-	/*
-	 * found private data in control, so copy it.
-	 * get possible component controls - get size of all pdata,
-	 * then memcpy with headers
-	 */
-	if (ipc_data_size) {
-		for (i = 0; i < widget->num_kcontrols; i++) {
-			memcpy(&process->data + offset,
-			       wdata[i].pdata->data,
-			       wdata[i].pdata->size);
-			offset += wdata[i].pdata->size;
-		}
-	}
-
-	process->size = ipc_data_size;
+	/* process data will be sent in the related kcontrol */
+	process->size = ipc_size;
 	swidget->private = process;
-err:
-	if (ret < 0)
-		kfree(process);
-out:
-	kfree(wdata);
+
```

Comment:
I really have a limited understanding of all this, it'd be good to clarify the commit message
a) what does 'streaming start' mean? open, hw_params, prepare, trigger? which of the phases will 'set the related kcontrol'
b) what does 'set the kcontrol' mean anyways? allocated, register?
c) what does 'memory duplication' mean? Does this mean we had some memory leak of something?

I find it really hard to figure out if this is a bug being corrected, an optimization, an improvement.

What was the issue that prompted you to look into this?

Reply:
I started looking into this when @singalsu was trying out beamformer and the widget was too big to fit to the ipc size -> error.

Anyway this piece of code was originally written by me to get the initial values of process components from its related kcontrol. So when the process comp is created -> copy the parameter data from kcontrol to comp. This was done because there was no any other nice way of setting the initial values (it is a long story, not able to put byte fields to widgets etc). This means also that initial kcontrol data is duplicated to the widget -> unnecessary memory usage. This also had limitation that if the copied kcontrol data (in the widget) exceeds max ipc size -> it is discarded as we don't have "large message" support in widget loading, only in kcontrol loading. Altogether this has many shortcomings... 

When dynamic pipelines features we're added, we are anyway always sending the kcontrol data to the related widget (earlier this was not the case) before there's anything processing related happening in the dsp -> code removed in this PR is unnecessary as the same data is anyway sent with kcontrol. If the code removed by this PR remains, the data is send twice.

Reply:
@juimonen if you are re-aligning the code to make use of a new flow enabled by dynamic pipelines, you really need to make a mention of the new flow in the commit message. You also want to mention what happens for static pipelines.

Reply:
@plbossart I changed the flow for static pipelines too to set up all widgets and send their kcontrol data after topology parsing is complete instead of during parsing. That change is what makes this code redundant.